### PR TITLE
support promises

### LINKED ISSUE #3: Promise interface
Add the capacity to return a promise from a run-async wrapped function.

### PR DESCRIPTION
**Pull request by jamestalmage**
closes #3 and #4

### CODE REVIEW

**Comment by jamestalmage** [comment]
hmm,
I'm using the `.catch()` method.
It has gotten to be pretty universal. Bluebird even calls `.then(resolveHandler, rejectHandler)` an anti-pattern.

Still `is-promise` only checks for a `then` method.

**Comment by jamestalmage** [comment]
Leaving as two commits. I can squash or remove the second one depending on which way you think we should go.

File: README.md [hunk 1]
```diff
@@ -38,11 +38,18 @@ var syncFn = function (a) {
   return 'running: ' + a;
 };
 
-runAsync(syncFn, function (answer) {
+runAsync(syncFn, function (err, answer) {
   console.log(answer); // 'running: sync'
 }, 'sync');
 ```
 
+Changes from 0.x.x
```
Comment (SBoudrias) [inline README.md]: I'll do the releases notes inside of the release page. No need to have them in the readme.

That being said, we'll need to update the readme to demo/explain the promise feature.

File: package.json [hunk 2]
```diff
@@ -22,9 +22,11 @@
   },
   "homepage": "https://github.com/SBoudrias/run-async",
   "dependencies": {
+    "is-promise": "^2.1.0",
     "once": "^1.3.0"
   },
   "devDependencies": {
+    "bluebird": "^2.10.2",
```
Comment (SBoudrias) [inline package.json]: Can we use a barebone es6 promise polyfill?

File: test.js [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 var assert = require('assert');
 var runAsync = require('./index');
+var Promise = require('bluebird');
 
 describe('runAsync', function () {
   it('run synchronous method', function (done) {
```
Comment (SBoudrias) [inline test.js]: We should have a test covering a failure case.

File: test.js [hunk 4]
```diff
@@ -47,7 +51,52 @@ describe('runAsync', function () {
...
+    runAsync(aFunc, function (err, val) {
+      assert.ifError(err);
+      done();
+    });
+  });
+
+  it('handles promises', function (done) {
+    var fn = function () {
+      return new Promise(function(resolve, reject) {
+        setTimeout(function () {
```
Comment (SBoudrias) [inline test.js]: Nit, we can use setImmediate here.

File: package.json (hunk 2 above)
Comment (jamestalmage) [inline package.json]: It's just a devDependency, so it won't impact anybody but people developing on it.

I tried `pinkie` at first, but it does not emit `unhandledRejection` events on process, so you had to hunt around for the error when it failed. Switching to `bluebird` surfaced errors immediately.

We could do [native-or-bluebird](https://www.npmjs.com/package/native-or-bluebird)

**Comment by SBoudrias** [comment]
What is the standard for rejected es6 promises? This is what we should base ourselves on.

Otherwise, code looks good, just had a couple comments.

**Comment by jamestalmage** [comment]
According [to this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), both are supported in es6.

I think the trending preferred style is to use `.catch`, but `.then(onResolve, onReject)` remains correct and has the advantage of being backwards compatible with whatever old promise implementation people throw at it. Also, in this particular case, it doesn't introduce ugly looking code since it is on a really compact line anyways (we aren't declaring the function inline).

I lean towards `.then(onResolve, onReject)`

**Comment by jamestalmage** [comment]
@SBoudrias 
OK, only two issues left are `catch` vs `then(.., onErr)`, and switching away from bluebird.

I think there are good reasons to leave it as is (see comments above), but I leave it up to you.

**Comment by SBoudrias** [comment]
I'm good with the `then`, just wanted to confirm as I haven't really followed the standardization of promises :tired_face: 

Thanks for following up on all this! It helps a lot

**Comment by jamestalmage** [comment]
what about bluebird?
Am I good to squash?

File: test.js (hunk 4 above)
Comment (SBoudrias) [inline test.js]: Oops, you missed that one ;)

**Comment by SBoudrias** [comment]
Okay with bluebird, I checked around for only a polyfill and it was more complex than necessary. Squash away!

**Comment by jamestalmage** [comment]
Done!

**Comment by jamestalmage** [comment]
>  I checked around for only a polyfill

I don't think there is such a thing. For having the tiniest little specification, they seem incredibly complex to implement correctly.

**Comment by SBoudrias** [comment]
I'll rework some documentation and publish a major version soon.

I'll also add you as a contributor to the project. You've been very helpful.

**Comment by jamestalmage** [comment]
Awesome! Thanks!

**Pull request merged**